include/aws/auth/private/xml_parser.h

```diff
@@ -0,0 +1,95 @@
+#ifndef AWS_AUTH_XML_PARSER
+#define AWS_AUTH_XML_PARSER
+
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include <aws/common/array_list.h>
+#include <aws/common/byte_buf.h>
+
+#include <aws/auth/exports.h>
+
+struct aws_xml_attribute {
+    struct aws_byte_cursor name;
+    struct aws_byte_cursor value;
+};
+
+struct aws_xml_node {
+    struct aws_byte_cursor name;
+    struct aws_array_list attributes;
+    struct aws_byte_cursor doc_at_body;
+    bool processed;
+};
+
+struct aws_xml_parser;
+
+/**
+ * Callback for when an xml node is encountered in the document. As a user you have a few options:
+ *
+ * 1. reject the document parsing at this point by returning false. This will immediately stop doc parsing.
+ * 2. call aws_xml_node_traverse() on the node to descend into the node with a new callback and user_data.
+ * 3. call aws_xml_node_as_body() to retrieve the contents of the node as text.
+ *
+ * return true to continue the parsing operation.
+ */
+typedef bool(
+    aws_xml_parser_on_node_encountered_fn)(struct aws_xml_parser *parser, struct aws_xml_node *node, void *user_data);
+
+struct aws_xml_parser {
+    struct aws_allocator *allocator;
+    struct aws_byte_cursor doc;
+    struct aws_array_list cb_stack;
+    struct aws_xml_attribute attributes[10];
+    struct aws_byte_cursor split_scratch[22];
+    bool stop_parsing;
+};
+
+/**
+ * Initialize the parser with xml document: doc.
+ */
+AWS_AUTH_API
+int aws_xml_parser_init(struct aws_xml_parser *parser, struct aws_allocator *allocator, struct aws_byte_cursor *doc);
```

Comment:
I feel like `doc` should be a param for `parse`. What if you want to reuse a parser later?

Reply:
currently it's tied to the doc for simplicity. the parser is mostly state for the doc it's working on, I feel like it would be awkward to change that.... we can always revisit if/when we decide to make this API public